---
title: Logistic Ordinal Regression of ROSMAP DLPFC Expression to NeuroPath Scores
output: html_notebook
---
Date of analysis update: `r date()`

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## Assumes the path to the NeuroPath_Regression folder is ~/Neuropath_Regression
## Load required libraries
library(data.table)
require(foreign)
require(ggplot2)
require(MASS)
require(Hmisc)
require(reshape2)
library(plyr)
library(tidyverse)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')
library(doParallel)
library(foreach)
library(synapser)

cl = makeCluster(detectCores()-1)
registerDoParallel(cl)
options(xtable.type="html")

dir_wd <- file.path("~", "NeuroPath_Regression")
dir_input <- file.path(dir_wd, "input_data")
if (!dir.exists(dir_input)) {
  dir.create(dir_input)
}

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

#### Pull Data
```{r PullData, eval=TRUE, message=TRUE, echo=TRUE, include=TRUE }
synLogin()

#Neuropath Scores
#SVA adjusted (Not Diagnosis Adjusted)
EXP <- read.table( synGet('syn21292032', downloadLocation=dir_input)$path, 
                   row.names=1, header=T, sep='\t' )

#Remove Leading X from column names
colnames(EXP) <- gsub( "X", "", colnames(EXP) )

Meta <- read.csv( synGet('syn3191087', downloadLocation=dir_input)$path, 
                  header=T, stringsAsFactors = F)
Meta2 <- read.csv( synGet('syn21323366', downloadLocation=dir_input)$path, 
                   header=T, stringsAsFactors = F )

ID_Key <- read.csv( synGet('syn3382527', downloadLocation=dir_input)$path )
ID_Key <- ID_Key[ ID_Key$rnaseq_data ==1, ]

#Filter for retained RNA-Seq IDs
ID_Key <- ID_Key[ ID_Key$rnaseq_id %in% colnames(EXP), ]
ID_Key <- ID_Key[ !duplicated(ID_Key$rnaseq_id), ]
row.names(ID_Key) <- ID_Key$projid

row.names(Meta) <- Meta$individualID
FullMeta <- cbind( Meta2, Meta[Meta2$individualID,] )
FullMeta <- FullMeta[ FullMeta$tissue == 'dorsolateral prefrontal cortex', ]
#FullMeta <- FullMeta[ FullMeta$notes == 'geneExpression (rnaSeq)', ]
FullMeta <- FullMeta[ FullMeta$assay == 'rnaSeq', ]

FullMeta <- FullMeta[ FullMeta$projid %in% ID_Key$projid, ]
FullMeta <- FullMeta[ !duplicated(FullMeta$individualID), ]

row.names(FullMeta) <- FullMeta$projid
row.names(ID_Key) <- ID_Key$projid
ID_Key <- ID_Key[row.names(FullMeta),]

FullMeta$RNASeqID <- ID_Key[row.names(FullMeta),]$rnaseq_id
row.names(FullMeta) <- FullMeta$RNASeqID

table( row.names(FullMeta) %in% colnames(EXP) )
table( colnames(EXP) %in% row.names(FullMeta) )

EXP <- t(EXP)[ row.names(FullMeta), ]
PartialMeta <- FullMeta[ ,c('ceradsc', 'braaksc', 'cogdx', 'dcfdx_lv' ) ]
table( row.names(PartialMeta) == row.names(EXP) )
```

#### Setup and Run Models
```{r SetAnaylsisParameters, eval=TRUE, message=TRUE, echo=TRUE, include=TRUE }
NeuroPath_Calc <- function( GN, Path, Exp ){
  #'@GN a character ENSG Gene name eg 'ENSG00000227232'
  #'@Path the character string of Neuropath column to use for model eg. EITHER: 'ceradsc', 'braaksc', 'cogdx', 'dcfdx_lv'
  #'@Exp a samples x genes matrix of expression
  #print(GN)
  Dat <- as.data.frame( cbind( PartialMeta, Gene = scale( as.numeric(Exp[,GN]) ) ), 
                        stringsAsFactors = F )
  Dat$ceradsc <- as.factor(Dat$ceradsc)
  Dat$braaksc <- as.factor(Dat$braaksc)
  Dat$cogdx <- as.factor(Dat$cogdx)
  Dat$dcfdx_lv <- as.factor(Dat$dcfdx_lv)
  
  formula <- as.formula(paste(Path, "Gene", sep = "~"))
  m <- polr(formula, data = Dat, Hess = TRUE)

  ctable <- coef(summary(m))
  
  PVal <- pnorm(abs(ctable["Gene", "t value"]), lower.tail = FALSE) * 2
  
  names( GN ) <- 'Gene'
  names( PVal ) <- 'PVal'
  OR <- exp(coef(m))
  names( OR ) <- 'OR'
  CI <- exp(confint(m))
  
  # TODO change to this
  #library(psych)
  #tmp6 = corr.test(x = Exp[,1:100], y = PartialMeta$braaksc, method = "spearman", adjust="fdr")
  
  return( c( GN, OR, CI, PVal ) )
}
```

```{r Runanalyses, eval=TRUE, message=TRUE, echo=TRUE, include=TRUE}
#Process Braak Scores - 113 failed
mark<-Sys.time()
Path <- 'braaksc'
BRAAK <- foreach( i=colnames(EXP), .export=c( 'Path', 'EXP', 'temp'), 
                  .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% {
                    NeuroPath_Calc( GN=i, Path=Path, Exp=EXP )
} 
Sys.time() - mark

#Process CERAD Scores - 17350 failed
mark<-Sys.time()
Path <- 'ceradsc'
CERAD <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), 
                  .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% {
                    NeuroPath_Calc( GN=temp[i], Path=Path, Exp=EXP )
}
Sys.time() - mark

#Process CogDX Scores - 1985 failed
mark<-Sys.time()
Path <- 'cogdx'
cogdx <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), 
                  .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% {
                    NeuroPath_Calc( GN=temp[i], Path=Path, Exp=EXP )
}
Sys.time() - mark

#Process DCFDX Level Scores - 1985 failed
mark<-Sys.time()
Path <- 'dcfdx_lv'
dcfdx <- foreach( i=1:length(colnames(EXP)), .export=c( 'Path', 'EXP', 'temp'), 
                  .packages = c('foreign', 'MASS', 'Hmisc'), .combine = 'rbind' ) %dopar% {
                    NeuroPath_Calc( GN=temp[i], Path=Path, Exp=EXP )
}
Sys.time() - mark

Cleaner <- function( DF, type ){
  DF <- as.data.frame( DF, stringsAsFactors=F )
  DF$OR <- as.numeric(DF$OR)
  DF$`2.5 %` <- as.numeric(DF$`2.5 %`)
  DF$`97.5 %` <- as.numeric(DF$`97.5 %`)
  DF$PVal <- as.numeric(DF$PVal)
  DF$PVal_Adj <- p.adjust( DF$PVal, method = 'fdr', n=dim(DF)[1] )
  DF$NeuroPath <- type
  return(DF)
}

BRAAK <- Cleaner( BRAAK, 'BRAAK' )
CERAD <- Cleaner( CERAD, 'CERAD' )
dcfdx <- Cleaner( dcfdx, 'DCFDX' )
cogdx <- Cleaner( cogdx, 'COGDX' )

Master <- as.data.frame( rbind( BRAAK,CERAD,dcfdx,cogdx ), stringsAsFactors=F )
Master <- Master[, c( 'Gene', 'OR', '2.5 %', '97.5 %', 'PVal', 'PVal_Adj', 'NeuroPath' ) ] 
colnames(Master) <- c( 'ENSG', 'OddsRatio', 'CI_Lower', 'CI_Upper', 'PVal', 
                       'PVal_Adj', 'NeuroPath_Type' ) 

print( paste0( 'Significantly associated genes by Neuropath Score: ' ))
table(Master[ Master$PVal_Adj < 0.05, ]$NeuroPath_Type) 

print( paste0( 'Frequency of Gene Across Significance to Neuropath Score: '))
table(table(Master[ Master$PVal_Adj < 0.05, ]$ENSG)) 

print( paste0( 'Significantly associated genes by Neuropath Score with an Odds Ratio < 1: ',
table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio < 1 )['TRUE'] ))

print( paste0( 'Significantly associated genes by Neuropath Score with an Odds Ratio > 1: ',
table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio > 1 )['TRUE'] ))

print( 'Significantly associated genes by Neuropath Score with an Odds Ratio > 1 by Score Type: ')
table(Master[ Master$PVal_Adj < 0.05, ]$OddsRatio > 1, 
      Master[ Master$PVal_Adj < 0.05, ]$NeuroPath_Type ) 

# Store results
#write.csv(Master, file = '~/NeuroPath_Regression/Neuropathology_Scores.csv',  row.names =F)
```

```{r GNAME}
library(biomaRt)
id.type = 'ensembl_gene_id'
host = 'https://jul2019.archive.ensembl.org'
  
message("Connecting to BioMart ...")
ensembl <- useMart("ENSEMBL_MART_ENSEMBL", host = host)
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)

Genes <- getBM(filters = id.type, attributes = c(id.type, 'hgnc_symbol'), 
               values = unique(Master$ENSG) , mart = ensembl)

Genes <- Genes[ !duplicated(Genes$ensembl_gene_id), ]
row.names(Genes) <- Genes$ensembl_gene_id

Master$GName <- Genes[Master$ENSG,]$hgnc_symbol

Master <- Master[ , c('ENSG', 'GName', 'OddsRatio', 'CI_Lower','CI_Upper', 
                      'PVal', 'PVal_Adj', 'NeuroPath_Type')
                ]

# Store results
write.csv(Master, file = file.path(dir_wd, "Neuropathology_Scores.csv"), row.names = F)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = "syn51769643" #'syn7525089';
activityName = 'ROSMAP DLPFC NeuroPath Regression Results';
activityDescription = 'Logistic Ordinal Regression of ROSMAP Gene Expression to Neuro Pathology scores (Braak, CERAD, CogDx, and DcfDx) ';
thisFileName <- 'Neuropath_Scores.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jaclynbeck-sage/NeuroPath_Regression", 
                             ref="branch", refName='main')
thisFile <- githubr::getPermlink(repository = thisRepo, 
                                 repositoryPath = paste0('code/', thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
activityName = 'ROSMAP DLPFC NeuroPath Regression Results';
activityDescription = 'Logistic Ordinal Regression of ROSMAP Gene Expression to Neuro Pathology scores (Braak, CERAD, CogDx, and DcfDx) ';

CODE <- synStore(Folder(name = "correlations_gene_neuropath", parent = parentId))

#Set Used SynIDs For Provenance
Syns_Used <- c( 'syn21292032', 'syn3191087', 'syn21323366', 'syn3382527' )

# Set annotations
all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'RNAseq',
  tissueTypeAbrv	= 'DLPFC', 
  study = 'ROSMAP', 
  organism = 'HomoSapiens',
  consortium	= 'ROSMAP',
  normalizationStatus	= TRUE,
  normalizationType	= 'CQN',
  rnaquantification = 'RSEM',
  genomeAssemblyID = 'GRCh38'
)

ENRICH_OBJ <- synStore( File( path = file.path(dir_wd, "Neuropathology_Scores.csv"), 
                              name = 'ROSMAP DLPFC NeuroPath Regression Results', 
                              parent = CODE$properties$id,
                              annotations = all.annotations), 
                        activityName = activityName, 
                        executed = c(Syns_Used, thisFile), 
                        activityDescription = activityDescription,
                        forceVersion = FALSE)
```

### R Source Code
[Github](`r thisFile`)

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
source(file.path(dir_wd, "utilityFunctions", "knitfile2synapseClient.R"))
source(file.path(dir_wd, "utilityFunctions", "hook_synapseMdSyntax_plot.R"))
createAndKnitToFolderEntityClient(file = file.path(dir_wd, "code", "Neuropath_Scores.Rmd"),
                                          parentId = "syn51769643", #"syn7525089",
                                          folderName = 'correlations_gene_neuropath')
```
